### PR TITLE
Rework the LocalStorage implementation

### PR DESCRIPTION
**Pull request by bdlukaa**
## What is the current behavior?

`shared_preferences` is currently used to store the user session

## What is the new behavior?

`hive` is now used.

## Aditional information

Fix for #8

Reworked the LocalStorage implementation. `LocalStorage` is now an abstract class that can be extended by others. Some classes are already built-in: 

- `HiveLocalStorage`, that implements Hive as the storage solution to persist the session on the device.
- `EmptyLocalStorage`, that disabled persistence.

An example of how to extend LocalStorage:

```dart
class EmptyLocalStorage extends LocalStorage {
  /// Creates a [LocalStorage] instance that disables persistance
  const EmptyLocalStorage()
      : super(
          initialize: _initialize,
          hasAccessToken: _hasAccessToken,
          accessToken: _accessToken,
          removePersistedSession: _removePersistedSession,
          persistSession: _persistSession,
        );

  static Future<void> _initialize() async {}
  static Future<bool> _hasAccessToken() => Future.value(false);
  static Future<String?> _accessToken() => Future.value(null);
  static Future<void> _removePersistedSession() async {}
  static Future<void> _persistSession(_) async {}
}
```

I also added a new function to `LocalStorage`: `initialize`. There you can initialize anything you need before performing any search operation. On the `HiveLocalStorage` implementation, it is used to open the box:

```dart
  static Future<void> _initialize() async {
    HiveCipher? encryptionCipher;
    if (encryptionKey != null) {
      encryptionCipher = HiveAesCipher(base64Url.decode(encryptionKey!));
    }
    await Hive.initFlutter('auth');
    await Hive.openBox(_hiveBoxName, encryptionCipher: encryptionCipher);
  }
```

It could be useful to improve fetching time for a lot of packages. We can take `shared_preferences` as an example:

```dart

late SharedPreferences _sharedPreferencesInstance;

class SharedPreferencesLocalStorage extends LocalStorage {
  const SharedPreferencesLocalStorage ()
      : super(
          initialize: _initialize,
          hasAccessToken: _hasAccessToken,
          accessToken: _accessToken,
          removePersistedSession: _removePersistedSession,
          persistSession: _persistSession,
        );

  static Future<void> _initialize() async {
    _sharedPreferencesInstance = await SharedPreferences.getInstance();
  }
  static Future<bool> _hasAccessToken() async {
    return _sharedPreferencesInstance.containsKey(...);
  }
  static Future<String?> _accessToken() {...}
  static Future<void> _removePersistedSession() async {...}
  static Future<void> _persistSession(_) async {...}
}
```

PS.: this was NOT tested in any app in production.

### CODE REVIEW

File: lib/src/supabase.dart [hunk 1]
```diff
@@ -52,7 +53,7 @@ class Supabase {
       _instance.log('***** Supabase init completed $_instance');
 
       await SupabaseAuth.initialize(
-        localStorage: localStorage ?? const LocalStorage(),
+        localStorage: localStorage ?? const HiveLocalStorage(),
```
Comment (atnegrete) [inline lib/src/supabase.dart:56]: I haven't got much dart 2.0 experience, but could the default `?? const HiveLocalStorage()` be moved up to the static initialize method parameter? 

My thought process here is, am I able to disable persistence or will it be forced with this package?

**Comment by atnegrete** [comment]
This is cool beans

File: lib/src/supabase.dart (hunk 1 above)
Comment (bdlukaa) [reply lib/src/supabase.dart:56]: you can create a custom LocalStorage implementation that does nothing. That way, it'll never persist the session. It's a great idea to have it implemented by default on the package

Comment (bdlukaa) [reply lib/src/supabase.dart:56]: I added the `EmptyLocalStorage` class to disable session persistance

File: lib/src/local_storage.dart [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+  /// The encryption key used by Hive. If null, the box is not encrypted
+  ///
+  /// This value should not be redefined in runtime, otherwise the user may
+  /// not be fetched correctly
+  ///
+  /// See also:
+  ///
+  ///   * <https://docs.hivedb.dev/#/advanced/encrypted_box?id=encrypted-box>
+  static String? encryptionKey;
```
Comment (bdlukaa) [inline lib/src/local_storage.dart:81]: How to add a encryption key?

```dart
HiveLocalStorage.encryptionKey = MY_ENCRYPTION_KEY;
```

See <https://docs.hivedb.dev/#/advanced/encrypted_box?id=encrypted-box> for more info on how to store your encryption key

**Comment by phamhieu** [comment]
I love the idea of `abstract LocalStorage class` and  `EmptyLocalStorage`. Nice work @bdlukaa 🚀🚀

> It could be useful to improve fetching time for a lot of packages

What do you mean `fetching time` here?

**Review by phamhieu — APPROVED** [review]

**Comment by bdlukaa** [comment]
> What do you mean fetching time here?

On shared preferences, the instance is requested at every function. Using initialize, we store the instance in a variable and use it when necessary

**Review by dshukertjr — APPROVED** [review]
Amazing work 🎉 LGTM!

**Comment by bdlukaa** [comment]
I'll merge after I do some testings

**Comment by bdlukaa** [comment]
Testings done! Everything works as expected

**Pull request merged**